src/System.Net.Security/src/System/Net/Security/Sslstream.Implementation.cs

```diff
@@ -159,256 +147,17 @@ internal void ValidateCreateContext(SslAuthenticationOptions sslAuthenticationOp
-                SslConnectionInfo info = Context.ConnectionInfo;
-                if (info == null)
-                {
-                    return 0;
-                }
-
-                return info.DataKeySize;
-            }
-        }
-
-        internal HashAlgorithmType HashAlgorithm
-        {
-            get
-            {
-                CheckThrow(true);
-                SslConnectionInfo info = Context.ConnectionInfo;
-                if (info == null)
-                {
-                    return (HashAlgorithmType)0;
-                }
-                return (HashAlgorithmType)info.DataHashAlg;
-            }
-        }
-
-        internal int HashStrength
-        {
-            get
-            {
-                CheckThrow(true);
-                SslConnectionInfo info = Context.ConnectionInfo;
-                if (info == null)
-                {
-                    return 0;
-                }
-
-                return info.DataHashKeySize;
-            }
-        }
-
-        internal ExchangeAlgorithmType KeyExchangeAlgorithm
-        {
-            get
-            {
-                CheckThrow(true);
-                SslConnectionInfo info = Context.ConnectionInfo;
-                if (info == null)
-                {
-                    return (ExchangeAlgorithmType)0;
-                }
-
-                return (ExchangeAlgorithmType)info.KeyExchangeAlg;
-            }
-        }
-
-        internal int KeyExchangeStrength
-        {
-            get
-            {
-                CheckThrow(true);
-                SslConnectionInfo info = Context.ConnectionInfo;
-                if (info == null)
-                {
-                    return 0;
-                }
-
-                return info.KeyExchKeySize;
-            }
-        }
-
-        internal SslProtocols SslProtocol
-        {
-            get
-            {
-                CheckThrow(true);
-                SslConnectionInfo info = Context.ConnectionInfo;
-                if (info == null)
-                {
-                    return SslProtocols.None;
-                }
-
-                SslProtocols proto = (SslProtocols)info.Protocol;
-                SslProtocols ret = SslProtocols.None;
-
-#pragma warning disable 0618 // Ssl2, Ssl3 are deprecated.
-                // Restore client/server bits so the result maps exactly on published constants.
-                if ((proto & SslProtocols.Ssl2) != 0)
-                {
-                    ret |= SslProtocols.Ssl2;
-                }
-
-                if ((proto & SslProtocols.Ssl3) != 0)
-                {
-                    ret |= SslProtocols.Ssl3;
-                }
-#pragma warning restore
+        private bool RemoteCertRequired => _context == null || _context.RemoteCertRequired;
 
-                if ((proto & SslProtocols.Tls) != 0)
-                {
-                    ret |= SslProtocols.Tls;
-                }
+        private object SyncLock => _context;
```

Comment:
What if _context is null?

Reply:
from my reading (I could be wrong) it is set during authentication, it is never then set to null. The lock is only used on post authentication messages (re-handshake for instance or during a key change). All methods post authentication that might touch it are protected with a CheckThrow(true).

Reply:
I see.  Can we rename the property then to be something like PostAuthenticateSyncLock, and add a Debug.Assert(IsAuthenticated) into the accessor?